src/benchmarks/micro/runtime/System.Reflection/Invoke.cs

```diff
@@ -0,0 +1,199 @@
+
+            s_ctor_int_string_struct_class = typeof(MyClass).
+                GetConstructor(new Type[] { typeof(int), typeof(string), typeof(MyBlittableStruct), typeof(MyClass) })!;
+
+            s_ctor_NoParams = typeof(MyClass).
+                GetConstructor(Array.Empty<Type>())!;
+
+            s_property_int = typeof(MyClass).
+                GetProperty(nameof(MyClass.I));
+
+            s_property_class = typeof(MyClass).
+                GetProperty(nameof(MyClass.O));
+
+            s_field_int = typeof(MyClass).
+                GetField(nameof(MyClass.i));
+
+            s_field_class = typeof(MyClass).
+                GetField(nameof(MyClass.o));
+        }
+
+        public static void Method_int_string_struct_class(int i, string s, MyBlittableStruct myStruct, MyClass myClass)
+        {
+            s_dummy++;
+        }
+
+        public static void Method_byref_int_string_struct_class(ref int i, ref string s, ref MyBlittableStruct myStruct, ref MyClass myClass)
+        {
+            s_dummy++;
+        }
+
+        public static void Method_nullableInt(int? i)
+        {
+            s_dummy++;
+        }
+
+        [Benchmark]
+        public void Method0_NoParms()
+        {
+            s_method.Invoke(s_MyClass, null);
+        }
+
+        [Benchmark]
+        public void StaticMethod4_int_string_struct_class()
+        {
+            object[] args = new object[] { 42, "Hello", default(MyBlittableStruct), s_MyClass };
+            s_method_int_string_struct_class.Invoke(null, args);
+        }
+
+        [Benchmark]
+        public void StaticMethod4_PreInstantiatedObjectArray_int_string_struct_class()
+        {
+            s_method_int_string_struct_class.Invoke(null, s_args);
+        }
+
+        [Benchmark]
+        public void StaticMethod4_ByRefParams_int_string_struct_class()
+        {
+            object[] args = new object[] { 42, "Hello", default(MyBlittableStruct), s_MyClass };
+            s_method_byref_int_string_struct_class.Invoke(null, args);
+        }
+
+        [Benchmark]
+        public void StaticMethod4_ByRefParams_PreInstantiatedObjectArray_int_string_struct_class()
+        {
+            s_method_byref_int_string_struct_class.Invoke(null, s_args);
+        }
+
+        [Benchmark]
+        public void StaticMethod1_NullableInt()
+        {
+            object[] args = new object[] { 42 };
+            s_method_nullableInt.Invoke(null, args);
+        }
+
+        [Benchmark]
+        public void Ctor0_NoParams()
+        {
+            s_ctor_NoParams.Invoke(null);
+        }
+
+        [Benchmark]
+        public void Ctor0_ActivatorCreateInstance_NoParams()
+        {
+            Activator.CreateInstance(typeof(MyClass));
+        }
+
+        [Benchmark]
+        public void Ctor4_int_string_struct_class()
+        {
+            object[] args = new object[] { 42, "Hello", default(MyBlittableStruct), s_MyClass };
```

Comment:
Would it be better to pull creating these arrays out of the benchmarks? They could presumably simply be reused.

Doesn't matter much as presumably it's very cheap but it'd be consistent.

Reply:
The benchmarks with "PreInstantiatedObjectArray" use a cached array. The rest (majority) allocate the array.

 I suppose I could flip it, and by default cache most.

Reply:
Why allocate each iteration for any of them though?

Reply:
It is measuring the actual real-world scenario where typically the arrays are not pre-allocated or pre-filled. Thus, these benchmarks are a fair, realistic benchmark especially when claiming that reflection is "X" times faster. This is also useful for when comparing against the proposed new reflection APIs being introduced in 8.0 which avoid the array allocation.

Given that, I do think I should flip the benchmarks so the default is pre-allocated (but not pre-filled) and then label the ones that allocate as "scenario"-based.